src/core/ParallelTaskManager.h

```diff
@@ -0,0 +1,364 @@
+  ParallelActionsInput myinput;
+/// This holds data for that the underlying action needs to do the calculation
+  D actiondata;
+/// This is used internally to gather the forces on the threads
+  void gatherThreads( ForceOutput& forces );
+public:
+  static void registerKeywords( Keywords& keys );
+  ParallelTaskManager(ActionWithVector* av);
+/// Setup the parallel task manager the three arguments are
+/// nind = the number of indices that are used per task
+/// nder = number of derivatives per component that is being calculated
+/// nt = number of derivatives that need to gathered over the threads (default of zero means all derivatives need to be calculated in a way that is thread safe)
+  void setupParallelTaskManager( std::size_t nind, std::size_t nder, std::size_t nt=0 );
+/// Copy the data from the underlying colvar into this parallel action
+  void setActionInput( const D& adata );
+/// Get the action input so we can use it
+  D& getActionInput();
+/// This runs all the tasks
+  void runAllTasks();
+/// Apply the forces on the parallel object
+  void applyForces( std::vector<double>& forcesForApply );
+};
+
+template <class T, class D>
+void ParallelTaskManager<T, D>::registerKeywords( Keywords& keys ) {
+  keys.addFlag("USEGPU",false,"run this calculation on the GPU");
+}
+
+template <class T, class D>
+ParallelTaskManager<T, D>::ParallelTaskManager(ActionWithVector* av):
+  action(av),
+  comm(av->comm),
+  ismatrix(false),
+  useacc(false),
+  nderivatives_per_component(0),
+  nthreaded_forces(0),
+  myinput(av->getPbc()) {
+  ActionWithMatrix* am=dynamic_cast<ActionWithMatrix*>(av);
+  if(am) {
+    ismatrix=true;
+  }
+  action->parseFlag("USEGPU",useacc);
+#ifdef __PLUMED_HAS_OPENACC
+  if( useacc ) {
+    action->log.printf("  using GPU to calculate this action\n");
+  }
+#else
+  if( useacc ) {
+    action->error("cannot use USEGPU flag as PLUMED has not been compiled with openacc");
+  }
+#endif
+}
+
+template <class T, class D>
+void ParallelTaskManager<T, D>::setupParallelTaskManager( std::size_t nind, std::size_t nder, std::size_t nt ) {
+  plumed_massert( action->getNumberOfComponents()>0, "there should be some components wen you setup the index list" );
+  std::size_t valuesize=(action->getConstPntrToComponent(0))->getNumberOfStoredValues();
+  for(unsigned i=1; i<action->getNumberOfComponents(); ++i) {
+    plumed_assert( valuesize==(action->getConstPntrToComponent(i))->getNumberOfStoredValues() );
+  }
+  myinput.ncomponents = action->getNumberOfComponents();
+  nderivatives_per_component = nder;
+  value_stash.resize( valuesize*action->getNumberOfComponents() );
+  myinput.nindices_per_task = nind;
+
+  if( nt==0 ) {
+    nthreaded_forces = action->getNumberOfDerivatives();
+  } else {
+    nthreaded_forces = nt;
+  }
+
+  unsigned t=OpenMP::getNumThreads();
+  if( useacc ) {
+    t = 1;
+  }
+  omp_forces.resize(t);
+  for(unsigned i=0; i<t; ++i) {
+    omp_forces.resize(nthreaded_forces);
```

Comment:
```suggestion
    omp_forces[i].resize(nthreaded_forces);
```
I think this is the original intention of this for loop